File: quote/src/database/TemplateSaver.vb
Imports System.Data.SqlClient
Imports DCS.Quote.Model.Template
Imports DCS.Quote.Model
Imports DCS.Quote.QuoteDataBase
Imports System.Reflection
Imports System.Data.OleDb
Imports System.Transactions
Imports DCS.Quote.QuoteDataBaseTableAdapters

Public Class TemplateSaver

    Public Function Save(ByVal q As Header) _
                        As Integer

        frmMain.frmMain.UseWaitCursor = True
        My.Application.DoEvents()

        ' Ensure the properies are updated
        frmMain.frmMain.Focus()

        Dim o As PrimaryPropeties = q.PrimaryProperties

        Dim newId As Integer
        Dim id As Integer = o.CommonID

        Dim adaptor As New QuoteDataBaseTableAdapters._QuoteTableAdapter
        If id > 0 Then
            adaptor.Connection.Open()
            o.CommonLastModified = Date.Now
            Dim test As Integer = adaptor.Update( _
                o.CustomerName, o.RequestForQuoteNumber, _
                o.PartNumber, False, Nothing, _
                o.CommonInitials, o.CommonCreatedDate, _
                o.LastModified, o.CommonID)
            adaptor.Connection.Close()
            newId = id
        Else
            adaptor.Connection.Open()
            adaptor.Transaction = adaptor.Connection.BeginTransaction
            adaptor.Insert(o.CustomerName, _
                o.PartNumber, o.RequestForQuoteNumber, False, Nothing, _
                o.CommonInitials, Date.Now, Date.Now)
            Dim cmd As OleDbCommand = New OleDbCommand( _
                "SELECT @@IDENTITY", adaptor.Connection)
            cmd.Transaction = adaptor.Transaction
            newId = CInt(cmd.ExecuteScalar())
            adaptor.Transaction.Commit()
            adaptor.Connection.Close()
            If id = 0 Then
                q.PrimaryProperties.SetID(newId)
            End If
        End If

        adaptor.Connection.Open()
        CommonSaver.DeleteProperties(newId)
        CommonSaver.SaveOtherProperties(newId, q.OtherProperties, False)
        CommonSaver.SaveComputationProperties(newId, q.ComputationProperties, False)
        CommonSaver.DeleteComponents(newId)
        CommonSaver.SaveComponents(q, newId, False)
        adaptor.Connection.Close()

        q.ComputationProperties.ClearDirty()
        q.OtherProperties.ClearDirty()
        q.PrimaryProperties.ClearDirty()
        q.ClearDirty()

        frmMain.frmMain.UseWaitCursor = False

        Return newId
    End Function

End Class

File: quote/Projects/PluginModelIO/TemplateSaver.cs
namespace Model.IO
{
    using System;
    using System.Data.SqlClient;
    using System.Globalization;
    using System.Linq;
    using System.Windows.Forms;
    
    using DB;
    
    public static class TemplateSaver
    {
        public static int Save(Common.Header header)
        {
            Cursor.Current = Cursors.WaitCursor;

            // Ensure the properies are updated
            Host.UI.UIApp.MainForm.Focus();

            var o = header.PrimaryProperties as Model.Template.PrimaryProperties;

            int newId = 0;
            int id = o.CommonId;

            var adaptor = new DB.QuoteDataBaseTableAdapters._QuoteTableAdapter();
            
            if (id > 0) 
            {
                adaptor.Connection.Open();
                o.CommonLastModified = System.DateTime.Now;
                
                adaptor.Update(
                    o.Customer.Name, 
                    o.RequestForQuoteNumber, 
                    o.PartNumber, 
                    false, 
                    null, 
                    o.CommonInitials, 
                    o.CommonCreatedDate, 
                    o.LastModified, 
                    o.Customer.Id, 
                    o.CommonId,
                    o.CommonId);
                
                adaptor.Connection.Close();
                newId = id;
            } 
            else 
            {
                adaptor.Connection.Open();
                var proxy = new QuoteTableProxy(adaptor);
                proxy.Transaction = adaptor.Connection.BeginTransaction();
                
                adaptor.Insert(
                    o.Customer.Name, 
                    o.RequestForQuoteNumber, 
                    o.PartNumber, 
                    false, 
                    0, 
                    o.CommonInitials, 
                    DateTime.Now, 
                    DateTime.Now, 
                    o.Customer.Id);
                
                var cmd = new SqlCommand("SELECT @@IDENTITY", adaptor.Connection);
                cmd.Transaction = proxy.Transaction;
                newId = Convert.ToInt32(cmd.ExecuteScalar(), CultureInfo.CurrentCulture);
                proxy.Transaction.Commit();
                adaptor.Connection.Close();
                
                if (id == 0) 
                {
                    header.PrimaryProperties.SetId(newId);
                }
            }

            adaptor.Connection.Open();
            CommonSaver.DeleteProperties(newId);
            CommonSaver.SaveNoteProperties(newId, header.NoteProperties);
            CommonSaver.SaveOtherProperties(newId, header.OtherProperties, true);
            CommonSaver.SaveComputationProperties(newId, header.ComputationProperties, true);
            CommonSaver.DeleteComponents(newId);
            CommonSaver.SaveComponents(header, newId, true);
            adaptor.Connection.Close();

            header.IsDirty = false;

            System.Windows.Forms.Cursor.Current = Cursors.Default;

            return newId;
        }
    }
}
